### DISCUSSION TITLE
Center of Buoyancy and the Model Water Level

### DISCUSSION THREAD

**Discussion by yman95** [BEMRosetta/BEMRosetta#20] [Q&A]
Hello,

Thanks for developing and updating BEMRosetta. I have been trying to learn and use NEMOH and WEC-Sim for my research, and this program has been a great help.

I have a few things I don't quite understand regarding the center of buoyancy and the center of motion. I made a simple sphere geometry to test out my WEC-Sim program. So, I assumed the geometry is made of aluminium alloy (rho=2710 kg/m^3) and is 20m in radius with thin shell (t=0.1m). I calculated the weight of the object (1.36E+8 N) and the required buoyancy force. Then how much of the volume is submerged. The depth of submersion is about 7.383m from the bottom of the sphere. Then my center of buoyancy should be the geometric center of the submerged portion which is calculated to be about 5.874m from the bottom. 

In BEMRosetta, is there a way to modify the location of CB? At the moment, it automatically determines the CB at about 4.867m (-2.517 in z from origin) from the bottom. I suspect that it did not take the weight of the object into consideration. CB didn't change even after my attempt to edit the mass of the object in the 'Process' tab under 'Mesh.' 

Then I get confused on the difference between the center of buoyancy and the center of motion. Doesn't the center of motion, CM, change with regard to external forces, such as the wave excitation force? How do I determine where to put CM in my Nemoh.cal file? 

Lastly, I modelled the geometry making the bottom 7.383m below the x-y plane (z<0), and this is reflected in the BEMRosetta to be considered as 'underwater.' Is this a good practice? How do you recommend in making a model to process in BEMRosetta to use in NEMOH and WEC-Sim?

Any help is greatly appreciated, and let me know if you need any additional files for verification. :)

[simple_sphere.zip](https://github.com/BEMRosetta/BEMRosetta/files/11009486/simple_sphere.zip)

**Comment by izabala123** [comment]
Hi yman94

I will try to answer some questions and please send me later more questions to go refining the asnwers and solving your doubts.

This is the body as it is, with the water line.
![image](https://user-images.githubusercontent.com/38589221/227192068-1c1fab28-072c-4738-aaab-a723664e94fb.png)

This is the submerged side.
![image](https://user-images.githubusercontent.com/38589221/227192133-fb435f01-f992-46aa-812e-c18108d7b652.png)

Entering the mass as 1.36e8/9.81, considering rho = 1025, g = 9.81, and clicking in Move/Archimede, the body is translated:
![image](https://user-images.githubusercontent.com/38589221/227192556-49f24f33-c381-4340-8055-da2a32014117.png)

CB is got automatically, with Archimede law. If you move the geometry, it will be recalculated.

The CM is set by hand or loaded. If the body is moved, the CM as well.

The centre of motions is the centre of calculation. For example, the added mass, excitacion force, stiffness matrix, etc, in roll, is different depending where de centre of motion is (abbreviated as C0).

For Wamit and Nemoh I would set the geometry so that draft is the 'real' draft with that water density.

I have answered some question but not all.  You can ask me more details.

Best regards
Iñaki

**Reply by yman95** [reply]
Thanks for replying, sorry for the late reply. I was away, and now I'm back.

For C0, as an initial input to NEMOH and WEC-Sim, do we consider it without any external forces, e.g., wave excitation force, mooring restrain? And in a simulation, if we consider wave latching, i.e., surge and pitch motion, the geometry and C0 will surely move. Is there a way to account for this or is this scenario not in the scope of BEM and is something WEC-Sim should handle?

And can I ask what do you mean by 'real draft'? Do you mean the geometry input for NEMOH? 

And a quick question on the BEM Solver - Nemoh v3 solver. For the mesh file, do we use the .stl file or the exported .dat file? 

Thanks again for the help,
Yi

**Comment by izabala123** [comment]
Hi Yi

C0 is only the centre of the system of coordinates, your system of reference. If you have got the potential flow coefficients referred to that C0, the inertia matrix has to be referred to it, and the time domain calculations too. If not, you will obtain wrong results for roll, pitch and yaw.

Don't worry about the movements of the C0 or the CG in the time domain simulations. The solver takes care of that. The C0, CG, etc have to be referred to the mesh in the position you created it. It is the same for the mooring fairleads. When you define the simulation, you set their x,y,z referred to the mesh. When the time domain simulation begins, they will move, but don't worry, the solver cares.

With 'real draft' I mean that you can set the mesh in a position respect to Z=0, so the mesh has a draft. However, the real draft will depend on Archimedes (body mass, mesh submerged volume and water density). For example, the 'real draft' of your body will be different if you use sea water or fresh water.

For Nemoh v3, as far as I know, you should have to convert the .STL to .DAT. I prefer to do it first to be sure that the conversion looks right.

If you have any doubts, as I final test to be sure that your inputs are right, I advice you to begin with even simpler geometries, and running them to see if the results are correct or not. I know that this focus is naive, but it usually works.

Best regards
Iñaki

**Comment by yman95** [comment]
Hello @izabala123 

I'm trying to continue with the real draft .stl file for NEMOH and WEC-Sim by using the Archimedes function in BEMRosetta.

The geometry has a radius of 5m, thickness of 0.1m, rho=2710 kg/m^3, so the mass will be 83445.7693 kg, as shown below:
![image](https://user-images.githubusercontent.com/16159470/231594482-9e2f5b1c-06b1-4fdc-842b-ec534004e98f.png)

After setting the mass and clicking on Archimedes, BEMRosetta shows me the following: 
![image](https://user-images.githubusercontent.com/16159470/231594604-0535f3d2-8342-4b05-a8de-1d9c86a1d3fd.png)

I figured I should use this .stl file as the geometry input for WEC-Sim, so I exported to stl.binary, but when I load in the exported .stl file, the center of gravity Cg is reset to the origin:
![image](https://user-images.githubusercontent.com/16159470/231594939-0bc05224-1553-4a43-91d0-d380737a6607.png)

I don't know how to read the exported .dat file for NEMOH, so I wonder if the center of gravity has been correctly recorded in the .dat file, and how do I export a version of the .stl file such that the Cg is at the correct position?

Maybe I manually set it in Nemoh.cal here?
![image](https://user-images.githubusercontent.com/16159470/231596226-f2caa3f7-dc1c-441f-a140-373792d1c3ef.png)


Thank,
Yi

**Comment by izabala123** [comment]
Hi Yi

Neither the .stl nor the .dat files formats save the centre of gravity.
Because of this, in the case of the Nemoh .dat, BEMRosetta tries to load
and save the 'Hydrostatics.dat' file, that includes the CG, when loading
and saving the .dat file with the mesh..
Does this answer your question?

Best regards
Iñaki

**Reply by yman95** [reply]
Got it. I'll make sure 'Hydrostatics.dat' has the correct Cg.

But any suggestions on where to put C0? For a floating geometry at equilibrium, I assume C0 will be the midpoint between Cg and Cb.

**Comment by izabala123** [comment]
Hi Yi

The importance of C0 lies in the fact that for some radiation-diffraction
solvers, and in time domain, the C0 must necessarily be in a specific
position, for example, at 0,0,0, or at the CG.
Apart from that, it is only an indication of the reference system for all
calculations. All calculations in the system must have the same C0.

Best regards
Iñaki

**Comment by yman95** [comment]
I will be testing for cases of C0 at origin and at CG.

Is it normal that when I export the .stl file into NEMOH's .dat file, the mesh seems to have missing pieces? shown below:
![image](https://user-images.githubusercontent.com/16159470/233188696-3dd45909-1003-465b-9b0c-ba87b0134d0b.png)

**Reply by yman95** [reply]
And there are cases where the missing slice is on the top of the mesh. Because of this, the displacement, and mass in Hydrostatics.dat are not consistent after running NEMOH's calculation. I assume the BEM calculations and the hydrodynamics will be off because of it.

For example, the mesh shown above have:
 XF =  -0.000 - XG =   0.000
 YF =   0.000 - YG =   0.000
 ZF =  -0.859 - ZG =   2.493
 Displacement =  0.7159592E+02
 Waterplane area =  0.5270277E+02
 Mass = 0.7338581E+05

but another run with the same procedure but the missing piece in .dat file is on the top side will have Hydrostatics.dat as
 XF =  -0.000 - XG =   0.000
 YF =   0.000 - YG =   0.000
 ZF =  -0.875 - ZG =   2.494
 Displacement =  0.8138976E+02
 Waterplane area =  0.5881902E+02
 Mass = 0.8342450E+05

The .dat file for the calculations above:
![image](https://user-images.githubusercontent.com/16159470/233190827-29338665-3042-43c5-9fac-c81ecb17c477.png)
![image](https://user-images.githubusercontent.com/16159470/233190913-eab354fe-b278-4f2d-8425-200f1c777651.png)

**Reply by yman95** [reply]
@izabala123 Also, is there a way to adjust the origin of the geometry without changing the Cg and Cb? I think the Translate function moves all of the points. I think WEC-Sim needs the origin to be at the Cg. The results from using the geometries shown above will result the model floating in the air above the water surface. Thanks again for your patience and help.

**Comment by izabala123** [comment]
Hi Yi

I cannot reproduce this weird behaviour using the simple_sphere.zip and simple_sphere.stl. Please send the .st file that produces this problem.

When you translate the mesh, the CG is moved, but the CB is set to the new position of the centre of buoyancy. 
If your time domain solver (WEC-Sim) requires the hydrodynamic coefficients to be referred to the CG, I would set, in the radiation-diffraction code (Nemoh), the C0 in the location of the CG. In Nemoh.cal, you can set, per each degree of freedom, the reference system in the CG. You don't need to move the mesh to a wrong location, the mesh has to be at the right Z height.

Best regards
Iñaki

**Reply by yman95** [reply]
Yes, sorry. I created a smaller sphere with r=5m. 
[Sphere_r5.zip](https://github.com/BEMRosetta/BEMRosetta/files/11286997/Sphere_r5.zip)

The calculated properties are in the .txt file.

**Comment by izabala123** [comment]
Hi Yi

Please download the last version, I have just upload it.
The problem has to be solved.

Best regards
Iñaki

**Reply by yman95** [reply]
After the update, I'm having trouble using the Archimede function to set the geometry to neutral buoyancy position using my calculated mass 83445.7693kg. Can you please confirm that it works on your end when using sphere_r5.stl file?
![image](https://user-images.githubusercontent.com/16159470/233865566-6d34b456-390c-43cc-869c-805f879fd80a.png)
![image](https://user-images.githubusercontent.com/16159470/233865576-a0f8201d-2829-452a-8293-8e42acf931b7.png)

**Comment by izabala123** [comment]
Hi Yi

Thank you for your report.
Please download the last version, I have just upload it.
The problem has to be solved.

Best regards
Iñaki

**Comment by yman95** [comment]
Hi @izabala123,

You are correct about putting C0 at the location of CG for NEMOH's BEM calculation. After working with WEC-Sim's developer, there is one issue left at this point which is the location of CG in the stl file.

Since CG is not saved in the stl and dat file, CG is defined at the origin which is at the water level. I believe NEMOH requires the origin to be at the water level and WEC-Sim requires it to be at CG for visualization. 

If I put CG at (0, 0, 2.494) in BEMRosetta, the final simulated geometry in WEC-Sim will float above water
![image](https://user-images.githubusercontent.com/16159470/236016395-3cc08f2f-8e05-40af-aab3-4c9114218561.png) ![image](https://user-images.githubusercontent.com/16159470/234712513-954e2b49-f3ff-4c74-9d41-cd003c2c98de.png)
Since the translate function moves the entire geometry, do you know if there is a way to move just the CG such that the geometry looks like 
![image](https://user-images.githubusercontent.com/16159470/231594604-0535f3d2-8342-4b05-a8de-1d9c86a1d3fd.png)
Perhaps I need to define an entirely new stl file just for WEC-Sim? 

Thanks,
Yi

**Comment by izabala123** [comment]
Hello Yi

I am sorry I don't understand you well. In Nemoh, you set in nemoh.cal the c0 per each degree of freedom. In addition, you can create the mesh .cal file. It is explained how to use it [here](https://lheea.ec-nantes.fr/research-impact/software-and-patents/eng-nemoh-mesh), and its output files. With it you can set the CG of the body, density and gravity. I don't know to use WEC-Sim, so I cannot help you about it. When you indicate that if you "put the CG at (0, 0, 2.494) in BEMRosetta, the final simulated geometry in WEC-Sim will float above water", I don't know what WEC-Sim is handling. For this reason, I would ask you to explain, in detail, the different steps so that I can help you, always taking into account my limitations with WEC_Sim.

Best regards
Iñaki

**Reply by yman95** [reply]
The calculation itself is correct now (for the linear case at the moment). There is only a small issue left which is the CG of the .stl geometry. 

WEC-Sim requires a .stl file as the geometry input, and for the linear case, it is purely for visualization. Only when I move to the nonlinear case does the .stl file mesh make differences in calculation. Here I used the saved stl file after the Archimedes function. e.g.,
![image](https://github.com/BEMRosetta/BEMRosetta/assets/16159470/bcdef5dc-1298-412c-9f00-28de41f98298)

The .stl geometry doesn't record CG, the saved geometry from BEMRosetta will have its CG at the water level instead of at the center, where it should be. 
![image](https://github.com/BEMRosetta/BEMRosetta/assets/16159470/c67a5a92-b38d-4498-be9e-8dccf49e3125)
So I just want to know if there is a way to move just the CG from water level to the center.

Let me know if it's still clear, thanks.

**Comment by izabala123** [comment]
I am sorry Yi, I don't understand you well.

If WEC-Sim requires a right CG for 2nd order calculations, and .stl doesn't
save the CG, how can we inform WEC-SIm about the position of the CG?

Iñaki

**Reply by yman95** [reply]
I think WEC-Sim takes the BEM results from NEMOH's calculation about CG, and that's how WEC-Sim know where the CG is. There is a difference in the requirement of where the CG should be in NEMOH and WEC-Sim. NEMOH has the origin at water level while WEC-Sim wants it to be at the CG. 

Can I refer you to my post in the WEC-Sim Github where he explained a bit on how WEC-Sim handles CG in his first and second response? https://github.com/WEC-Sim/WEC-Sim/issues/1044

I believe the simulation/calculation should be correct. It seems like I just need to have a stl geometry with the origin at its CG. The workaround I have in mind is produce a separate stl file as the WEC-Sim input. 

Hopefully this is clearer,
Yi

**Comment by izabala123** [comment]
Hello Yi

As far I understand, you could move the mesh in BR to the right position,
and save it.
Then, you set in the nemoh files the cg in the right position.
In this way, WEC-Sim will be happy.
As you know how WEC-Sim behaves, you can do it perfectly.

Best regards
Iñaki

**Comment by yman95** [comment]
Hi @izabala123,

I would like to report a bug in generating .dat file when a cube mesh is rotated 45 degrees in yaw. The diagonal mesh on the bottom face of the cube is not generated properly.
![image](https://github.com/BEMRosetta/BEMRosetta/assets/16159470/8780b8e9-fb01-4a99-be86-89a416a919b0)

The stl file and the WEC-Sim example .dat file are attached.

Thank you,
Yi

[Cube.zip](https://github.com/BEMRosetta/BEMRosetta/files/11682213/Cube.zip)

**Reply by yman95** [reply]
And it seems like it is the same case for cylinder as well
![image](https://github.com/BEMRosetta/BEMRosetta/assets/16159470/32a575d2-826d-4e39-b84d-bc2c10b16012)
[WEC-Sim_example_cylinder.zip](https://github.com/BEMRosetta/BEMRosetta/files/11683428/WEC-Sim_example_cylinder.zip)

Regards,
Yi

**Comment by izabala123** [comment]
Dear Yi

I have opened your files, and I see that rotating the files doesn't create any problem.
The _sub file has that problem, but from the file, without rotating it.

![image](https://github.com/BEMRosetta/BEMRosetta/assets/38589221/73df1d4e-8e01-4e38-b907-f663f5920e3e)

Please indicate me more simply the problem. Thank you.

Best regards
Iñaki

**Reply by yman95** [reply]
Sorry, let me clarify. It's saving the rotated stl file into NEMOH's .dat format will cause the inconsistencies in the mesh. The stl files themselves are fine. When I try to save the rotated .stl file into .dat format, those _sub files will have the mesh not complete. Thank you :)

**Comment by izabala123** [comment]
Thank you Yi, you got something

Yes, there is something in the .stl that disturbs BEMRosetta. The simplest way to avoid the problem now is to disable the symmetry when saving, unclicking the option like in the picture. Let me some days to solve the real problem.
![image](https://github.com/BEMRosetta/BEMRosetta/assets/38589221/c3e51157-4871-462d-84ac-7cfbf1859968)
Best regards
Iñaki

**Reply by yman95** [reply]
Unchecking the box fixes the problem. Thanks for your time and effort.

**Comment by izabala123** [comment]
Thanks to you for noticing the details and letting me know. 
It's not difficult to make a tool that works in general, but without help like yours, it's impossible to make it work every time.
When I can, I will solve this problem and let you know.

Thank you Yi
Iñaki
